examples/tc_port_whitelist/src/bpf/tc.bpf.c

```diff
@@ -0,0 +1,256 @@
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_endian.h>
+
+//#include <linux/if_ether.h>
+#define ETH_P_IP    0x0800
+#define ETH_P_IPV6  0x86DD
+
+//#include <linux/pkt_cls.h>
+#define TC_ACT_UNSPEC	(-1)
+#define TC_ACT_OK		0
+#define TC_ACT_RECLASSIFY	1
+#define TC_ACT_SHOT		2
+#define TC_ACT_PIPE		3
+#define TC_ACT_STOLEN		4
+#define TC_ACT_QUEUED		5
+#define TC_ACT_REPEAT		6
+#define TC_ACT_REDIRECT		7
+#define TC_ACT_TRAP		8
+
+u8 rc_allow = TC_ACT_UNSPEC;
+u8 rc_disallow = TC_ACT_SHOT;
+
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __uint(max_entries, 10);
+    __type(value, u16);
+    __type(key, u32);
+} ports SEC(".maps");
+
+/*
+ * Determine an ip4 header
+ *
+ * @param eth, ethernet header
+ * @param data_end, end of the entire packet
+ *
+ * @return a struct iphdr if this is an ipv4 packet
+ *         NULL otherwise
+ */
+static struct iphdr* is_ipv4(struct ethhdr *eth, void *data_end) 
+{
+    struct iphdr *iph = NULL;
+    if (!eth || !data_end) {
+        return NULL;
+    }
+
+    if ((void*)eth + sizeof(*eth) + sizeof(*iph) > data_end) {
+        return NULL;
+    }
+    
+    if (eth->h_proto == bpf_htons(ETH_P_IP)) {
+        iph = (struct iphdr*)((void*)eth + sizeof(*eth));
+    }
+    return iph;
+}
+
+/*
+ * Determine an ip6 header
+ *
+ * @param eth, ethernet header
+ * @param data_end, end of the entire packet
+ *
+ * @return a struct ipv6hdr if this is an ipv6 packet
+ *         NULL otherwise
+ */
+struct ipv6hdr* is_ipv6(struct ethhdr *eth, void *data_end) 
+{
+    struct ipv6hdr *iph = NULL;
+    if (!eth || !data_end) {
+        return NULL;
+    }
+
+    if ((void*)eth + sizeof(*eth) + sizeof(*iph) > data_end) {
+        return NULL;
+    }
+    
+    if (eth->h_proto == bpf_htons(ETH_P_IPV6)) {
+        iph = (struct ipv6hdr*)((void*)eth + sizeof(*eth));
+    }
+    return iph;
+}
+
+/*
+ * Determine udp header 
+ *
+ * @param iph, iphdr or ipv6hdr depending on hdr_size
+ * @param hdr_sz, size of the iph.
+ *        valid values are either sizeof(struct iphdr) or sizeof(ipv6hdr)
+ * @param data_end, end of the entire packet
+ *
+ * @return a struct udphdr if this is an udp packet
+ *         NULL otherwise
+ */
+struct udphdr* is_udp(void *iph, u8 hdr_sz, void *data_end)
+{
+    struct udphdr *udph = NULL;
+    if (!iph || !data_end) {
+        return NULL;
+    }
+
+    if ((void*)(iph + hdr_sz + sizeof(*udph)) > data_end) {
+        return NULL;
+    }
+
+    int proto = -1;
+    if (hdr_sz == sizeof(struct iphdr)) {
+        struct iphdr *v4 = (struct iphdr*)iph;
+        proto = v4->protocol;
+    } else if (hdr_sz == sizeof(struct ipv6hdr)) {
+        struct ipv6hdr *v6 = (struct ipv6hdr*)iph;
+        proto = v6->nexthdr;
+    }
+
+    if (proto == IPPROTO_UDP) {
+        udph = (struct udphdr*)((void*)iph + hdr_sz);
```

Comment:
I don't think this is correct - ipv4 headers are variable size lengths so in the ipv4 case, the tcp or udp header would be at `iph + (iph->ihl * 4)`